docs: add guide for rewrites (#948)
ory/docs: docs/keto/guides/userset-rewrites.mdx

---
title: Userset Rewrites
---

import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
import CodeFromRemote from "@theme/CodeFromRemote"
const sha = "fc39b65f8be9b23bbb4e19480706964d61505c5e"
const file = (path) => `https://github.com/ory/keto/blob/${sha}/contrib/rewrites-example/${path}`

This guide will explain how to configure your your namespaces and relations using the
[Ory Permission Language](../reference/ory-permission-language).

The example describes a file store. Individual files are organized in a folder hierarchy, and can be accessed by users or groups
of users. Through the Ory Permission Language we can specify that if a user has access to a folder, the user also has access to
all files in that folder.

## Setup and Configuration

First, [install Keto](../install.mdx). Next, you need the following `keto.yaml` (at `contrib/rewrites-example/keto.yaml`):

<CodeFromRemote src={file("keto.yaml")} />

In the _namespaces_ key, you see the line `config: file://./namespaces.keto.ts`. This needs to point to your namespace
configuration in the Ory Permission Language.

> :::info Backwards compatibility

Author: I'm not sure what you mean. How would it look like?

> :::info Backwards compatibility

Reviewer: `:::info Backwards compatibility` 
vs
`:::info` 

If you add some text after the admonition denominator (tip/info/note/warning), the text you added becomes the title of this admonition, if you will. I don't think I explained it well enough so maybe try it locally and see what's the difference between the two ;)

Namespace configurations without the Ory Permission Language are still supported by either specifying the namespaces directly or a
"naked" URI, e.g. `namespaces: file://.namespaces.yaml`.

:::

Next, we create a file with our namespace configuration as follows:

<CodeFromRemote src={file("namespaces.keto.ts")} />

:::tip Editor support

If you are using a text editor with TypeScript support, you can get extra help when using the Ory Permission Language. Just make
sure that the file `contrib/rewrites-examples/lib.ts` is in the same folder as the file you are editing. It contains all
definitions to type-check the config.

In the future, we will make the lib file available at an easy-to-import location.

:::

### Starting Ory Keto

After you created both configuration files (keto.yaml and namespaces.keto.ts), you can start Ory Keto with:

```sh
$ keto serve --config ./path/to/keto.yaml
```

## Creating the tuples

> Now that Ory Keto is running, we can create the tuples either through the Keto CLI.

Reviewer: I think you wanted to say that tuples can be created through the CLI or by loading files, right? The sentence looks unfinished the way it is now ;)

Group           developer               members         patrik
Group           developer               members         User:Patrik
Group           developer               members         User:Henning
Folder          keto/                   viewers         Group:developer#members
File            keto/README.md          parents         Folder:keto/
Folder          keto/src/               parents         Folder:keto/
File            keto/src/main.go        parents         Folder:keto/src/
File            private                 owners          User:Henning
```

## Checking for permissions

Now, let's check some permissions! Some queries to try:

```
$ keto check User:Patrik view File keto/src/main.go
Allowed
```

Patrik can view "keto/src/main.go", because it is in the folder "keto/src", which is in the folder "keto", which has as "viewers"
the group "developer", of which Patrik is a member of.

```
$ keto check User:Patrik view File private
Denied
```

Patrik cannot view the private file, since that file is not part of any folder hierarchy Patrik has access to.

```
$ keto check User:Henning view File private
Allowed

$ keto check User:Henning edit File private
Allowed
```

> Henning can both edit and view the private file, since he is an "owner" of it.

Author: Bullets didn't look good, but I made some subheaders.

> id: userset-rewrites title: Define file access permissions with userset rewrites sidebar_label: Userset rewrites

Reviewer 2: Please just fix it so we can get this done it just takes literally 2 seconds to do it:

```suggestion
id: userset-rewrites
title: Define file access permissions with userset rewrites
sidebar_label: Userset rewrites
```

---
id: userset-rewrites
title: Guide to Userset Rewrites
sidebar_label: Userset Rewrites
---

import Tabs from "@theme/Tabs"
import TabItem from "@theme/TabItem"
import CodeFromRemote from "@theme/CodeFromRemote"
const sha = "fc39b65f8be9b23bbb4e19480706964d61505c5e"
const file = (path) => `https://github.com/ory/keto/blob/${sha}/contrib/rewrites-example/${path}`

This guide will explain how to configure your your namespaces and relations using the
[Ory Permission Language](../reference/ory-permission-language).

The example describes a file store. Individual files are organized in a folder hierarchy, and can be accessed by users or groups

> of users. Through the Ory Permission Language we can specify that if a user has access to a folder, the user also has access to
> all files in that folder.

Reviewer: ```suggestion
of users. Using the Ory Permission Language you can specify that if a user has access to a folder, the user also has access to
all files in that folder.
```

## Setup and Configuration

> First, [install Keto](../install.mdx). Next, you need the following `keto.yaml` (at `contrib/rewrites-example/keto.yaml`):

Reviewer: ```suggestion
First, [install Keto](../install.mdx). Next, create the `keto.yaml` file and save it at the `contrib/rewrites-example/` path:
```

<CodeFromRemote src={file("keto.yaml")} />

> In the namespaces key, you see the line `config: file://./namespaces.keto.ts`. This needs to point to your namespace configuration

Reviewer: ```suggestion
In the `namespaces` key, you see the line `config: file://./namespaces.keto.ts`. This needs to point to your namespace configuration
```

in the Ory Permission Language.

:::info

Namespace configurations without the Ory Permission Language are still supported by either specifying the namespaces directly or a
"naked" URI, e.g. `namespaces: file://.namespaces.yaml`.

:::

> Next, we create a file with our namespace configuration as follows:

Reviewer: ```suggestion
Next, create a file with the namespace configuration:
```

<CodeFromRemote src={file("namespaces.keto.ts")} />

> :::tip Editor support

Reviewer: ```suggestion
:::tip
```

> If you are using a text editor with TypeScript support, you can get extra help when using the Ory Permission Language. Just make

Reviewer: ```suggestion
If you are using a text editor with TypeScript support, you can get extra help when using the Ory Permission Language. Make
```

sure that the file `contrib/rewrites-examples/lib.ts` is in the same folder as the file you are editing. It contains all
definitions to type-check the config.

:::

### Starting Ory Keto

> After you created both configuration files (`keto.yaml` and `namespaces.keto.ts`), you can start Ory Keto with:

Reviewer: ```suggestion
After you created both configuration files (`keto.yaml` and `namespaces.keto.ts`), run this command to start Ory Keto:
```

```sh
$ keto serve --config ./path/to/keto.yaml
```

## Creating the tuples

> Now that Ory Keto is running, we can create the tuples through the Keto CLI.

Reviewer: ```suggestion
Now that Ory Keto is running, create relation tuples using the Keto CLI.
```

> We use the following relation tuples to showcase the namespace configuration. In short, it sets up a "developer" group with two

Reviewer: ```suggestion
The following relation tuples showcase the namespace configuration. In short, it sets up a "developer" group with two
```

members, and a folder hierarchy. Through the rules in the Ory Permission Language, every member of the "developer" group can
access the files in the hierarchy.

> As usual, it is still possible to create additional fine-grained permission rules for certain objects, as is done for the
> "private" file.

Reviewer: ```suggestion
You can create additional fine-grained permission rules for certain objects, similar to the "private" file.
```

<CodeFromRemote src={file("relation-tuples/tuples.json")} />

To load the file into Ory Keto, run this command:

```sh
$ keto relation-tuple create ./contrib/rewrites-example/relation-tuples/tuples.json

NAMESPACE       OBJECT                  RELATION NAME   SUBJECT
Group           developer               members         patrik
Group           developer               members         User:Patrik
Group           developer               members         User:Henning
Folder          keto/                   viewers         Group:developer#members
File            keto/README.md          parents         Folder:keto/
Folder          keto/src/               parents         Folder:keto/
File            keto/src/main.go        parents         Folder:keto/src/
File            private                 owners          User:Henning
```

## Checking for permissions

Now, let's check some permissions! Some queries to try:

### Transitive permissions for objects in the hierarchy

> Patrik can view "keto/src/main.go", because it is in the folder "keto/src", which is in the folder "keto", which has as "viewers"
> the group "developer", of which Patrik is a member of.

Reviewer: ```suggestion
Patrik can view `keto/src/main.go`. This file is in the `keto/src` folder, which is in `keto`.  The `keto` directory has the "developer" group as its "viewers".  Patrik is a member of the "developer" group.
```
